File: go/topjslibs/google_search_result.go
package topjslibs

import "regexp"

var searchResultLink = regexp.MustCompile(`<a href="([^"]+?)"[^>]+?data-ved="[^"]+?"[^>]+?onmousedown="[^"]+?"><br>`)

type GoogleSearchResult struct {
	linksLimit int
	fetchHtml  func() string
}

func (googleSearchResult *GoogleSearchResult) getLinks() []string {
	html := googleSearchResult.fetchHtml()
	links := searchResultLink.FindAllStringSubmatch(html, -1)
	result := make([]string, 0, googleSearchResult.linksLimit)

	for _, regexGroups := range links {
		if len(regexGroups) < 2 {
			continue
		}

		if len(result) == googleSearchResult.linksLimit {
			break
		}

		result = append(result, regexGroups[1])
	}

	return result
}

File: rust/src/top_js_libs/google_search_result.rs
use regex::Regex;

lazy_static! {
    static ref SEARCH_RESULT_LINK: Regex = Regex::new(r#"<a href="([^"]+?)"[^>]+?data-ved="[^"]+?"[^>]+?onmousedown="[^"]+?"><br>"#).unwrap();
}

pub struct GoogleSearchResult<'a> {
    links_limit: usize,
    fetch_html: Box<dyn Fn() -> String + 'a>,
}

impl<'a> GoogleSearchResult<'a> {
    pub fn new<'b>(links_limit: usize, fetch_html: Box<dyn Fn() -> String + 'b>) -> GoogleSearchResult {
        GoogleSearchResult {
            links_limit,
            fetch_html,
        }
    }

    pub fn get_links(&self) -> Vec<String> {
        let html = (self.fetch_html)();
        let mut result = Vec::new();

        for capture in SEARCH_RESULT_LINK.captures_iter(&html) {
            if capture.len() < 2 {
                continue;
            }

            if result.len() == self.links_limit {
                break;
            }

            result.push(capture[1].to_string());
        }

        result
    }
}